File: AI/codes/llm-train-python/model/TransformerBlock.py
import torch.nn as nn
from .MultiHeadAttention import MultiHeadAttention
from .FeedForward import FeedForward
from .LayerNorm import LayerNorm

class TransformerBlock(nn.Module):
    def __init__(self, cfg):
        super().__init__()
        self.att = MultiHeadAttention(
            d_in=cfg['emb_dim'],
            d_out=cfg['emb_dim'],
            context_length=cfg['context_length'],
            num_heads=cfg['n_heads'],
            dropout=cfg['drop_rate'],
            qkv_bias=cfg['qkv_bias'],
        )
        self.feedforward = FeedForward(cfg)
        self.norm1 = LayerNorm(cfg['emb_dim'])
        self.norm2 = LayerNorm(cfg['emb_dim'])
        self.drop_shortcut = nn.Dropout(cfg['drop_rate'])

    def forward(self, x):
        # 注意力模块中的快捷连接
        shortcut = x
        x = self.norm1(x)
        x = self.att(x)
        x = self.drop_shortcut(x)
        x = x + shortcut

        # 前馈网络中的快捷连接
        shortcut = x
        x = self.norm2(x)
        x = self.feedforward(x)
        x = self.drop_shortcut(x)

        # 将原始的输入加回到输出中
        x = x + shortcut
        return x

File: AI/codes/llm-train-cpp/src/model/TransformerBlock.cpp
#include "llm/model/TransformerBlock.hpp"

#include "llm/ops.hpp"

namespace llm {

TransformerBlock::TransformerBlock(const GPTConfig& cfg)
    : att(cfg.emb_dim, cfg.emb_dim, cfg.context_length, cfg.n_heads, cfg.qkv_bias, cfg.device),
      ff(cfg), norm1(cfg.emb_dim, cfg.device), norm2(cfg.emb_dim, cfg.device) {}

// 带残差连接的 Transformer 块：x: (B, T, C) -> 返回: (B, T, C)（C=emb_dim）
Tensor TransformerBlock::forward(const Tensor& x) {
    Tensor y = att.forward(norm1.forward(x)); // norm1+注意力，(B,T,C)->(B,T,C)
    Tensor a = ops::add(x, y);                 // 残差相加，(B,T,C)
    Tensor z = ff.forward(norm2.forward(a));   // norm2+前馈，(B,T,C)->(B,T,C)
    return ops::add(a, z);                     // 残差相加，(B,T,C)
}

std::vector<Tensor*> TransformerBlock::parameters() {
    auto p = att.parameters();
    auto append = [&](std::vector<Tensor*> q) {
        p.insert(p.end(), q.begin(), q.end());
    };
    append(ff.parameters());
    append(norm1.parameters());
    append(norm2.parameters());
    return p;
}

} // namespace llm
